### PR TITLE
Remove IAssemblyProvider

### PR DESCRIPTION
**Pull request by javiercn**
* Removes IAssemblyProvider and their implementations from the code base.
* Adds DefaultAssemblyDiscoveryProvider to perform assembly discovery during startup and add the assemblies to the list of discovered parts.
* Updates AddMvcDnx to find the assemblies using the library manager and add them to the list of application parts.

### CODE REVIEW

**Comment by javiercn** [comment]
@pranavkm @rynowak

File: src/Microsoft.AspNetCore.Mvc.Core/ApplicationParts/DefaultAssemblyPartDiscoveryProvider.cs [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+using System.Reflection;
+using Microsoft.Extensions.DependencyModel;
+
+namespace Microsoft.AspNetCore.Mvc.ApplicationParts
+{
+    // Discovers assemblies that are part of the MVC application using the DependencyContext.
+    internal static class DefaultAssemblyPartDiscoveryProvider
```
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.Core/ApplicationParts/DefaultAssemblyPartDiscoveryProvider.cs]: Is this intentionally `internal`?

**Comment by pranavkm** [comment]
Could we move the Provider to an `.Internal` namespace instead? Doesn't seem like there's anything in particular about it that would require it + methods in it to be `internal`.

Code looks :shipit:

File: src/Microsoft.AspNetCore.Mvc.Core/ApplicationParts/DefaultAssemblyPartDiscoveryProvider.cs (hunk 1 above)
Comment (javiercn) [inline src/Microsoft.AspNetCore.Mvc.Core/ApplicationParts/DefaultAssemblyPartDiscoveryProvider.cs]: It is, @rynowak

Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/ApplicationParts/DefaultAssemblyPartDiscoveryProvider.cs]: This should be `.Internal` namespace not `internal static class`

**Comment by javiercn** [comment]
@rynowak updated

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/DefaultAssemblyPartDiscoveryProvider.cs [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+using Microsoft.Extensions.DependencyModel;
+
+namespace Microsoft.AspNetCore.Mvc.Internal
+{
+    // Discovers assemblies that are part of the MVC application using the DependencyContext.
+    public static class DefaultAssemblyPartDiscoveryProvider
+    {
+        private const string NativeImageSufix = ".ni";
+
+        internal static HashSet<string> ReferenceAssemblies { get; } = new HashSet<string>(StringComparer.Ordinal)
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/DefaultAssemblyPartDiscoveryProvider.cs]: These should be public or private. We don't really use the `internal` keyword

**Comment by rynowak** [comment]
:shipit:

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/DefaultAssemblyPartDiscoveryProvider.cs (hunk 2 above)
Comment (javiercn) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/DefaultAssemblyPartDiscoveryProvider.cs]: They are internal for unit testing

Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/DefaultAssemblyPartDiscoveryProvider.cs]: Nope. At least one of these methods is called by product code, it should be public.

**Comment by PinpointTownes** [comment]
Thanks for keeping the MVC/DNX package updated :clap:

**Pull request closed without merging**